File: dotnet/src/SemanticKernel/CoreSkills/MathSkill.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Globalization;
using System.Threading.Tasks;
using Microsoft.SemanticKernel.Orchestration;
using Microsoft.SemanticKernel.SkillDefinition;

namespace Microsoft.SemanticKernel.CoreSkills;

/// <summary>
/// MathSkill provides a set of functions to make Math calculations.
/// </summary>
/// <example>
/// Usage: kernel.ImportSkill("math", new MathSkill());
/// Examples:
/// {{math.Add}}         => Returns the sum of FirstNumber and SecondNumber (provided in the SKContext)
/// </example>
public class MathSkill
{
    /// <summary>
    /// Returns the Addition result of initial and amount values provided.
    /// </summary>
    /// <param name="initialValueText">Initial value as string to add the specified amount</param>
    /// <param name="context">Contains the context to get the numbers from</param>
    /// <returns>The resulting sum as a string.</returns>
    [SKFunction("Adds value to a value")]
    [SKFunctionName("Add")]
    [SKFunctionInput(Description = "The value to add")]
    [SKFunctionContextParameter(Name = "Amount", Description = "Amount to add")]
    public Task<string> AddAsync(string initialValueText, SKContext context) =>
        AddOrSubtractAsync(initialValueText, context, add: true);

    /// <summary>
    /// Returns the Sum of two SKContext numbers provided.
    /// </summary>
    /// <param name="initialValueText">Initial value as string to subtract the specified amount</param>
    /// <param name="context">Contains the context to get the numbers from</param>
    /// <returns>The resulting subtraction as a string.</returns>
    [SKFunction("Subtracts value to a value")]
    [SKFunctionName("Subtract")]
    [SKFunctionInput(Description = "The value to subtract")]
    [SKFunctionContextParameter(Name = "Amount", Description = "Amount to subtract")]
    public Task<string> SubtractAsync(string initialValueText, SKContext context) =>
        AddOrSubtractAsync(initialValueText, context, add: false);

    private static Task<string> AddOrSubtractAsync(string initialValueText, SKContext context, bool add)
    {
        if (!int.TryParse(initialValueText, NumberStyles.Any, CultureInfo.InvariantCulture, out var initialValue))
        {
            return Task.FromException<string>(new ArgumentOutOfRangeException(nameof(initialValueText), initialValueText, "Initial value provided is not in numeric format"));
        }

        string contextAmount = context["Amount"];
        if (!int.TryParse(contextAmount, NumberStyles.Any, CultureInfo.InvariantCulture, out var amount))
        {
            return Task.FromException<string>(new ArgumentOutOfRangeException(nameof(context), contextAmount, "Context amount provided is not in numeric format"));
        }

        var result = add ?
            initialValue + amount :
            initialValue - amount;

        return Task.FromResult(result.ToString(CultureInfo.InvariantCulture));
    }
}

File: java/plugins/semantickernel-plugin-core/src/main/java/com/microsoft/semantickernel/coreskills/MathSkill.java
// Copyright (c) Microsoft. All rights reserved.
package com.microsoft.semantickernel.coreskills;

import com.microsoft.semantickernel.skilldefinition.annotations.DefineSKFunction;
import com.microsoft.semantickernel.skilldefinition.annotations.SKFunctionInputAttribute;
import com.microsoft.semantickernel.skilldefinition.annotations.SKFunctionParameters;

import java.math.BigDecimal;

/**
 * MathSkill provides a set of functions to make Math calculations.
 *
 * <ul>
 *   <li><code>input</code> is the first number in the operation.
 *   <li><code>amount</code> is the second number in the operation.
 * </ul>
 *
 * <p>Usage:
 *
 * <blockquote>
 *
 * <pre>
 * kernel.ImportSkill("math", new MathSkill());
 *
 * {{math.add}}         => Returns the sum of input and amount
 * {{math.subtract}}    => Returns the differentce of input from amount
 * </pre>
 *
 * </blockquote>
 *
 * <p>This skill uses BigDecimal to perform the number conversion and the calculations.
 */
public class MathSkill {

    /**
     * Returns the addition result of input and amount values provided.
     *
     * <p>This skill attempts to convert the numbers to BigDecimal and then add them.
     *
     * <p>If a conversion error occurs, the skill will throw an exception.
     *
     * @param input Initial value as string to add the specified amount
     * @param amount The amount to be added to input
     * @return The resulting sum as a String.
     */
    @DefineSKFunction(description = "Adds amount to a value.", name = "add")
    public String add(
            @SKFunctionInputAttribute(description = "The value to add to.") String input,
            @SKFunctionParameters(name = "amount", description = "The amount to be added to value.")
                    String amount) {

        BigDecimal bValue = new BigDecimal(input);

        return bValue.add(new BigDecimal(amount)).toString();
    }

    /**
     * Returns the subtraction result of input and amount values provided.
     *
     * <p>This skill attempts to convert the numbers to BigDecimal and then subtract them.
     *
     * <p>If a conversion error occurs, the skill will throw an exception.
     *
     * @param input Initial value as string to subtract the specified amount
     * @param amount The amount to be subtracted from input
     * @return The resulting difference as a String.
     */
    @DefineSKFunction(description = "Subtracts amount from value.", name = "Subtract")
    public String subtract(
            @SKFunctionInputAttribute(description = "The value to subtract from.") String input,
            @SKFunctionParameters(
                            name = "amount",
                            description = "The amount to be subtracted from value.")
                    String amount) {

        BigDecimal bValue = new BigDecimal(input);

        return bValue.subtract(new BigDecimal(amount)).toString();
    }
}
